File: providers/dm/lib/dm/parse.ex
defmodule Dm.Parse do
  @moduledoc """
  Parses a command line into positionals and options.

  Options are `--flag value`, `--flag "value"`, or a bare `--flag`.

  A quoted value that is a JSON string literal, as the `{{key|json}}`
  template filter renders, is decoded. That keeps quotes, newlines, and
  ` --words` inside a `--body` intact. A quoted value that is not a JSON
  string literal runs to the last quote before the next ` --flag` or the end
  of input, so hand-typed quotes inside a value still survive.

  Positionals end at the first `--flag` that stands at a token boundary. A
  positional that is a JSON string literal, as `{{key|json}}` renders, is
  decoded first, so ` --words` and newlines inside an `append` text or a
  `search` query survive.

  Empty values, the bare literal `null` (the `json` filter's rendering of an
  absent argument), and the literal "false" for booleans count as absent. The
  literal "true", as a template renders a set boolean, parses as a bare flag.

  The scanner walks the line once and never backtracks over a value, so a
  multi-megabyte `--base64` or `--body` parses in linear time. A lazy regex
  over the value hits the PCRE match limit on such input and drops the option.
  """

  @flag_re ~r/(?:^|\s)--([a-z][a-z0-9-]*)/
  @after_quote_re ~r/^(?:\s+--[a-z][a-z0-9-]*(?:\s|$)|\s*$)/
  @after_token_re ~r/^(?:\s|$)/
  @token_re ~r/^(?:"((?:[^"\\]|\\.)*)"|'([^']*)'|(\S+))/s

  @doc """
  Splits a request message into the command line and the request body.

  The command line is the first line. The request body is everything after
  the first newline, or `nil` when the message is a single line.
  """
  @spec split(String.t()) :: {String.t(), String.t() | nil}
  def split(message) when is_binary(message) do
    case String.split(message, "\n", parts: 2) do
      [header, body] -> {header, body}
      [header] -> {header, nil}
    end
  end

  @spec parse(String.t()) :: {[String.t()], %{String.t() => String.t() | true}}
  def parse(line) when is_binary(line) do
    {args, rest} = positionals(String.trim(line), [])
    {Enum.reverse(args), scan_opts(rest, %{})}
  end

  # Positionals run from the start of the line to the first `--flag` that
  # stands at a token boundary. A positional that is a JSON string literal is
  # decoded, so a ` --word` inside it cannot end the positionals. Other
  # positionals follow shell quoting.
  defp positionals(text, acc) do
    text = String.trim_leading(text)

    case text do
      "" ->
        {acc, ""}

      <<"--", c, _::binary>> when c in ?a..?z ->
        {acc, text}

      <<?", _::binary>> ->
        case json_string(text, @after_token_re) do
          {:ok, value, rest} -> positionals(rest, [value | acc])
          :error -> shell_token(text, acc)
        end

      _ ->
        shell_token(text, acc)
    end
  end

  defp shell_token(text, acc) do
    [matched | groups] = Regex.run(@token_re, text)

    value =
      case groups do
        [dq | _] when dq != "" -> unescape(dq)
        [_, sq | _] when sq != "" -> sq
        [_, _, bare] -> bare
        _ -> ""
      end

    positionals(after_offset(text, byte_size(matched)), [value | acc])
  end

  defp scan_opts(text, acc) do
    case Regex.run(@flag_re, text, return: :index) do
      [_, {start, len}] ->
        key = binary_part(text, start, len)
        {value, rest} = take_value(after_offset(text, start + len))
        scan_opts(rest, put_opt(acc, key, value))

      nil ->
        acc
    end
  end

  # The value follows at least one whitespace character. A quoted value that
  # is a JSON string literal ending where a ` --flag` or the end of input
  # follows is decoded. Any other quoted value ends at the first quote that a
  # ` --flag` or the end of input follows. Anything else that does not start
  # with `--` is one bare token.
  defp take_value(rest) do
    case Regex.run(~r/^\s+/, rest, return: :index) do
      [{0, ws}] -> take_value(after_offset(rest, ws), rest)
      nil -> {nil, rest}
    end
  end

  defp take_value("", rest), do: {nil, rest}
  defp take_value("--" <> _, rest), do: {nil, rest}

  defp take_value(<<?", _::binary>> = text, rest) do
    case json_string(text, @after_quote_re) do
      {:ok, value, after_literal} ->
        {{:quoted, value}, after_literal}

      :error ->
        case close_quote(text) do
          {:ok, close} ->
            {{:quoted, binary_part(text, 1, close - 1)}, after_offset(text, close + 1)}

          :error ->
            bare(text, rest)
        end
    end
  end

  defp take_value(text, rest), do: bare(text, rest)

  defp bare(text, rest) do
    case Regex.run(~r/^\S+/, text, return: :index) do
      [{0, len}] -> {{:bare, binary_part(text, 0, len)}, after_offset(text, len)}
      nil -> {nil, rest}
    end
  end

  defp json_string(text, boundary_re) do
    with {:ok, len} <- json_string_length(text),
         after_literal = after_offset(text, len),
         true <- Regex.match?(boundary_re, after_literal),
         {:ok, value} <- decode_json_string(binary_part(text, 0, len)) do
      {:ok, value, after_literal}
    else
      _ -> :error
    end
  end

  # Byte length of the JSON string literal at the start of the binary,
  # quotes included. Escapes and quotes are ASCII, so scanning bytes is safe.
  defp json_string_length(<<?", rest::binary>>), do: json_string_end(rest, 1)

  defp json_string_end(<<?\\, _, rest::binary>>, n), do: json_string_end(rest, n + 2)
  defp json_string_end(<<?", _::binary>>, n), do: {:ok, n + 1}
  defp json_string_end(<<_, rest::binary>>, n), do: json_string_end(rest, n + 1)
  defp json_string_end(<<>>, _n), do: :error

  defp decode_json_string(literal) do
    case :json.decode(literal) do
      value when is_binary(value) -> {:ok, value}
      _ -> :error
    end
  rescue
    _ -> :error
  end

  defp close_quote(text) do
    text
    |> :binary.matches("\"")
    |> Enum.drop(1)
    |> Enum.find_value(:error, fn {pos, 1} ->
      if Regex.match?(@after_quote_re, after_offset(text, pos + 1)), do: {:ok, pos}
    end)
  end

  defp after_offset(bin, offset), do: binary_part(bin, offset, byte_size(bin) - offset)

  defp put_opt(acc, key, value) do
    key = String.replace(key, "-", "_")

    case value do
      nil -> Map.put(acc, key, true)
      {:quoted, ""} -> acc
      {:quoted, "false"} -> acc
      {:bare, "false"} -> acc
      {:quoted, "true"} -> Map.put(acc, key, true)
      {:bare, "true"} -> Map.put(acc, key, true)
      {:bare, "null"} -> acc
      {_, v} -> Map.put(acc, key, v)
    end
  end

  defp unescape(s), do: String.replace(s, ~r/\\(.)/, "\\1")
end

File: go/cmd/dm-provider/parse.go
package main

import (
	"encoding/json"
	"regexp"
	"strings"
)

// The parser reads one command line into positionals and options.
//
// An option is "--flag value", "--flag \"value\"", or a bare "--flag".
//
// A quoted value that is a JSON string literal, as the {{key|json}} filter of
// the CLI writes, is decoded. Quotes, newlines and " --words" inside a
// --body therefore survive. Any other quoted value runs to the last quote
// before the next " --flag", so quotes that a person typed survive as well.
//
// The positionals end at the first "--flag" that stands at a token boundary.
//
// An empty value, the literal null, and the literal false count as absent.
// The literal true reads as a bare flag.
//
// The scanner walks the line once, so a body of several megabytes parses in
// one pass.

var (
	flagPattern       = regexp.MustCompile(`(^|\s)--([a-z][a-z0-9-]*)`)
	afterQuotePattern = regexp.MustCompile(`^(\s+--[a-z][a-z0-9-]*(\s|$)|\s*$)`)
	afterTokenPattern = regexp.MustCompile(`^(\s|$)`)
	spacePattern      = regexp.MustCompile(`^\s+`)
	barePattern       = regexp.MustCompile(`^\S+`)
	tokenPattern      = regexp.MustCompile(`^("((?:[^"\\]|\\.)*)"|'([^']*)'|(\S+))`)
)

// splitMessage splits a request into its command line and its body. The body
// is everything after the first newline.
func splitMessage(message string) (line string, body string, hasBody bool) {
	if index := strings.Index(message, "\n"); index >= 0 {
		return message[:index], message[index+1:], true
	}
	return message, "", false
}

// parseLine reads one command line.
func parseLine(line string) ([]string, map[string]any) {
	args, rest := positionals(strings.TrimSpace(line), nil)
	return args, scanOptions(rest, map[string]any{})
}

// positionals runs from the start of the line to the first flag that stands
// at a token boundary.
func positionals(text string, found []string) ([]string, string) {
	for {
		text = strings.TrimLeft(text, " \t\n\r")

		switch {
		case text == "":
			return found, ""

		case strings.HasPrefix(text, "--") && len(text) > 2 && text[2] >= 'a' && text[2] <= 'z':
			return found, text

		case strings.HasPrefix(text, `"`):
			if value, rest, ok := jsonString(text, afterTokenPattern); ok {
				found = append(found, value)
				text = rest
				continue
			}
			fallthrough

		default:
			value, rest := shellToken(text)
			found = append(found, value)
			text = rest
		}
	}
}

// shellToken reads one token: a double quoted value, a single quoted value,
// or a run without space.
func shellToken(text string) (string, string) {
	match := tokenPattern.FindStringSubmatchIndex(text)
	if match == nil {
		return "", ""
	}

	groups := tokenPattern.FindStringSubmatch(text)
	value := ""

	switch {
	case groups[2] != "":
		value = unescape(groups[2])
	case groups[3] != "":
		value = groups[3]
	default:
		value = groups[4]
	}
	return value, text[match[1]:]
}

func scanOptions(text string, found map[string]any) map[string]any {
	for {
		match := flagPattern.FindStringSubmatchIndex(text)
		if match == nil {
			return found
		}

		name := text[match[4]:match[5]]
		value, rest := takeValue(text[match[5]:])
		putOption(found, name, value)
		text = rest
	}
}

// value is what follows a flag: nothing, a quoted run, or one bare token.
type value struct {
	text   string
	quoted bool
	absent bool
}

// takeValue reads the value of one flag. The value follows at least one
// space.
func takeValue(rest string) (value, string) {
	space := spacePattern.FindString(rest)
	if space == "" {
		return value{absent: true}, rest
	}

	text := rest[len(space):]

	switch {
	case text == "", strings.HasPrefix(text, "--"):
		return value{absent: true}, rest

	case strings.HasPrefix(text, `"`):
		if found, after, ok := jsonString(text, afterQuotePattern); ok {
			return value{text: found, quoted: true}, after
		}
		if close, ok := closeQuote(text); ok {
			return value{text: text[1:close], quoted: true}, text[close+1:]
		}
		return bareValue(text, rest)

	default:
		return bareValue(text, rest)
	}
}

func bareValue(text, rest string) (value, string) {
	found := barePattern.FindString(text)
	if found == "" {
		return value{absent: true}, rest
	}
	return value{text: found}, text[len(found):]
}

// jsonString reads a JSON string literal that ends at a boundary.
func jsonString(text string, boundary *regexp.Regexp) (string, string, bool) {
	length, ok := jsonStringLength(text)
	if !ok {
		return "", "", false
	}

	after := text[length:]
	if !boundary.MatchString(after) {
		return "", "", false
	}

	var found string
	if err := json.Unmarshal([]byte(text[:length]), &found); err != nil {
		return "", "", false
	}
	return found, after, true
}

// jsonStringLength is the length of the JSON string literal at the front,
// quotes included. Escapes and quotes are ASCII, so reading bytes is safe.
func jsonStringLength(text string) (int, bool) {
	if !strings.HasPrefix(text, `"`) {
		return 0, false
	}

	for index := 1; index < len(text); {
		switch text[index] {
		case '\\':
			index += 2
		case '"':
			return index + 1, true
		default:
			index++
		}
	}
	return 0, false
}

// closeQuote finds the quote that a flag or the end of the line follows.
func closeQuote(text string) (int, bool) {
	for index := 1; index < len(text); index++ {
		if text[index] == '"' && afterQuotePattern.MatchString(text[index+1:]) {
			return index, true
		}
	}
	return 0, false
}

func putOption(found map[string]any, name string, given value) {
	key := strings.ReplaceAll(name, "-", "_")

	switch {
	case given.absent:
		found[key] = true
	case given.text == "" && given.quoted:
		return
	case given.text == "false":
		return
	case given.text == "true":
		found[key] = true
	case given.text == "null" && !given.quoted:
		return
	default:
		found[key] = given.text
	}
}

// unescape drops the backslash of every escaped character.
func unescape(text string) string {
	var out strings.Builder
	for index := 0; index < len(text); index++ {
		if text[index] == '\\' && index+1 < len(text) {
			index++
		}
		out.WriteByte(text[index])
	}
	return out.String()
}

// option reads one option as a string. A bare flag is not a string.
func option(options map[string]any, name string) (string, bool) {
	text, ok := options[name].(string)
	return text, ok
}

// flagSet says whether a bare flag stands.
func flagSet(options map[string]any, name string) bool {
	set, ok := options[name].(bool)
	return ok && set
}
